src/lib/mdns/MdnsCache.h

```diff
@@ -0,0 +1,183 @@
+    }
+
+    // insert this entry into the cache.
+    // return error if cache is full
+    // TODO:   have an eviction policy so if the cache is full, an entry may be deleted.
+    //         One policy may be Least-time-to-live
+    CHIP_ERROR Insert(PeerId peerId, const Inet::IPAddress & addr, uint16_t port, Inet::InterfaceId iface, uint32_t TTLms)
+    {
+        const uint64_t currentTime = mTimeSource.GetCurrentMonotonicTimeMs();
+
+        MdnsCacheEntry * entry;
+
+        entry = FindPeerId(peerId, currentTime);
+        if (entry)
+        {
+            // update timeout if found entry
+            entry->expiryTime = currentTime + TTLms;
+            entry->TTL        = TTLms; // in case it changes */
+            return CHIP_NO_ERROR;
+        }
+
+        VerifyOrReturnError(entry = findSlot(currentTime), CHIP_ERROR_TOO_MANY_KEYS);
+
+        // have a free slot for this entry
+        entry->peerId     = peerId;
+        entry->ipAddr     = addr;
+        entry->port       = port;
+        entry->ifaceId    = iface;
+        entry->TTL        = TTLms;
+        entry->expiryTime = currentTime + TTLms;
+        elementsUsed++;
+
+        return CHIP_NO_ERROR;
+    }
+
+    CHIP_ERROR Delete(PeerId peerId)
+    {
+        MdnsCacheEntry * pentry;
+        const uint64_t currentTime = mTimeSource.GetCurrentMonotonicTimeMs();
+
+        VerifyOrReturnError(pentry = FindPeerId(peerId, currentTime), CHIP_ERROR_KEY_NOT_FOUND);
+
+        MarkEntryUnused(*pentry);
+        return CHIP_NO_ERROR;
+    }
+
+    // given a peerId, find the parameters if its in the cache, or return error
+    CHIP_ERROR Lookup(PeerId peerId, Inet::IPAddress & addr, uint16_t & port, Inet::InterfaceId & iface)
+    {
+        MdnsCacheEntry * pentry;
+        const uint64_t currentTime = mTimeSource.GetCurrentMonotonicTimeMs();
+
+        VerifyOrReturnError(pentry = FindPeerId(peerId, currentTime), CHIP_ERROR_KEY_NOT_FOUND);
+
+        addr  = pentry->ipAddr;
+        port  = pentry->port;
+        iface = pentry->ifaceId;
+
+        return CHIP_NO_ERROR;
+    }
+
+    // only useful if MDNS_LOGGING is set.   If not used, should be optimized out
+    void DumpCache()
+    {
+        int i = 0;
+
+        MdnsLogProgress(Discovery, "cache size = %d", elementsUsed);
+        for (MdnsCacheEntry & e : mLookupTable)
+        {
+            if (e.peerId == nullPeerId)
+            {
+                MdnsLogProgress(Discovery, "Entry %d unused", i);
+            }
+            else
+            {
+                char address[100];
+
+                e.ipAddr.ToString(address, sizeof address);
+                MdnsLogProgress(Discovery, "Entry %d: node %lx fabric %lx, port = %d, address = %s", i, e.peerId.GetNodeId(),
+                                e.peerId.GetFabricId(), e.port, address);
+            }
+            i++;
+        }
+    }
+
+private:
+    struct MdnsCacheEntry
+    {
+        PeerId peerId;
```

Comment:
Use of the `PeerId` object implies that all advertising nodes will have non-colliding service names (likely, but not guaranteed), and also suffers from the fact that `PeerId` is currently incomplete and does not provide unambiguous node identification because it includes neither fabric index nor root CA public key.

This also is non-spec-compliant because it implies nodes advertise fabricId-nodeId, when we know that the spec mandates they advertise fabricReference-nodeId. Why don't we just write this module in a spec-compliant way now?  Nothing is stopping us, and this incremental approach doesn't add value.

Reply:
The spec says they advertise "compressed Fabric ID"  -- what exactly is a "fabric index"?

This is from the spec:
4.2.2.1. Operational Instance Name

For CHIP operational discovery the DNS‑SD instance name is constructed from a 64-bit compressed Fabric identifier, and a 64-bit Node identifier, as assigned by the commissioner, each expressed as a fixed-length sixteen-character hexadecimal string, encoded as ASCII (UTF-8) text using capital letters, separated by a hyphen. For example, a CHIP Node with CHIP compressed fabric identifier 2906-C908-D115-D362 and CHIP Node identifier 8FC7-7724-01CD-0696 has CHIP operational discovery DNS‑SD instance name 2906C908D115D362-8FC7772401CD0696.

Reply:
See "4.2.2.2. Compressed Fabric Identifier" which defines what that thing is and how it's related to a fabric id.

Reply:
This was a previous PR #7411 for the MDNS cache with this [comment stream ](https://github.com/project-chip/connectedhomeip/pull/7411#discussion_r658231090) which indicated use peerid for now.

Reply:
> The spec says they advertise "compressed Fabric ID" -- what exactly is a "fabric index"?

Fabric index is a concept from the operational credentials cluster.  Clients can interact with the fabrics to which a cluster server is commissioned by referencing them by index.

> This was a previous PR #7411 for the MDNS cache with this comment stream which indicated use peerid for now.

Yes, I am raising a potential objection to that.

My fundamental question is this: given the fact that fully-qualified root-CA public key / fabric ID / node ID tuples are not guaranteed to be collision-free for operational advertisement with fabric reference, how should such collisions?  I can imagine these possible approaches:

* this actually isn't a problem and in the theoretical case where a node is joined to two or more fabrics where fabric references collide, nodes should simply collapse these colliding advertisements into a single advertisement; in this approach, we can pass `peerId` or `uint64_t fabricReference, uint64_t nodeId` to the advertisement interface
* OR we can't or do not wish to tolerate colliding fabric references and we will reject such advertisement from the mDNS code; such an error would have to propagate back to the commissionee's onboarding logic and would have to prevent operational-channel rendezvous, which would force both commissioner and commissionee to fail back to the commissioning channel; in this scenario we must pass `peerId` to the advertisement interface so that mDNS can positively identify collisions
* OR we can't or do not wish to tolerate colliding fabric references; therefore, we must call this out as an error from the commissionee when they first receive new, colliding operational credentials; in this scenario, we should not pass `peerId` and should instead pass `uint64_t fabricReference, uint64_t nodeId` to make it clear that the mDNS code is not responsible for identifying collisions

Do we know which of these three approaches we need to take?  That would tell us whether passing peerId really is appropriate.  And if we can choose not to do so, we have an opportunity to immediately change this code to provide this spec-compliant interface: `advertise(uint64_t fabricReference, uint64_t nodeId)`.  Callers would be free to still pass `uint64_t fabricId` until the rest of the stack catches up.

Reply:
I vote for:

> this actually isn't a problem and in the theoretical case where a node is joined to two or more fabrics where fabric references collide, nodes should simply collapse these colliding advertisements into a single advertisement; in this approach, we can pass peerId or uint64_t fabricReference, uint64_t nodeId to the advertisement interface

Because of the nature of the compressed fabric ID generation, the odds are *beyond* realistic for a collision to be found on a local network, unless explicitly being attempted by an attacker, who could in any case have done worse by just learning of an existing service, and "taking it over" on next network reboot by trying to advertise it faster than the legitimate one, such that the legitimate one would "lose" during mDNS probing.

The overall odds are such that I consider this problem a non-problem. If a node is joined to 2 fabrics that collapse to the same compressed ID, *and* the node ID is the same, I'd argue it's fine to advertise just one time, since it will be findable. CASE relies on other means to identify the peers, and CASE will succeed.

Reply:
Thanks all for the conversation.  I'm in agreement with the approach that uses peerId.  This makes fine sense now and in the future.  And waiting to refactor to fabricReference-nodeId when that is rolled out more generally in the code is a perfectly reasonable way to sequence work.